numpy/core/src/multiarray/nditer_pywrap.c

```diff
@@ -2331,6 +2395,45 @@ npyiter_ass_subscript(NewNpyArrayIterObject *self, PyObject *op,
     return -1;
 }
 
+static PyObject *
+npyiter_enter(NewNpyArrayIterObject *self)
+{
+    if (self->iter == NULL) {
+        PyErr_SetString(PyExc_RuntimeError, "operation on non-initialized iterator");
+        return NULL;
+    }
+    if (self->managed == CONTEXT_EXITED) {
+        PyErr_SetString(PyExc_ValueError, "cannot reuse iterator after exit");
+        return NULL;
+    }
+    self->managed = CONTEXT_INSIDE;
```

Comment:
Should probably also check that you're not re-entering the iterator

Reply:
Either that, or allow re-entrant `with`s, but only cleanup on the outermost one.

Reply:
Reentrant is OK (test added), multiple cleanups are harmless. Once the cleanup is finished, we set `self->managed = CONTEXT_EXITED` so reentering a closed nditer fails (there is a test for that in `test_nditer.py`, line 2847

Reply:
I guess the question is, is this the behavior we expect?
```
with it:
    with it:
        do_stuff_with_it()  # ok
    do_stuff_with(it)  # error: already closed
```

It does seem to match how `contextlib.closing` works, so I suppose is consistent.